### PR TITLE
stat/distuv: fix Laplace.Fit

### LINKED ISSUE #638: stat/distuv/Laplace Fit panics when weights is nil
### What are you trying to do?
According to the documentation 
```golang
// If weights is nil, then all the weights are 1.
func (l *Laplace) Fit(samples, weights []float64) 
```

### What did you do?

However the following code panics
```golang
package main

import (
		"gonum.org/v1/gonum/stat/distuv"

)

func main() {
	var d distuv.Laplace
	d.Fit([]float64{1}, nil)
}
```

### What did you expect to happen?

Nothing
### What actually happened?
`panic: distuv: slice length mismatch`

### What version of Go and Gonum are you using?
go version go1.11.1 darwin/amd64
513db5a0a3aa4236f0ace2604a014c157876c32e

### Does this issue reproduce with the current master?

Yes

What is the intended behavior nil should or should not be accepted. I can submit a patch if required.

### LINKED ISSUE #639: stat/distuv/Laplace Fit not properly estimating scaling factor
### What are you trying to do?
Want to fit a Laplace distribution to a set of sample data
According to the documentation 
```golang
// Fit sets the parameters of the probability distribution from the
// data samples x with relative weights w.
func (l *Laplace) Fit(samples, weights []float64) {
```

### What did you do?

```golang
package main

import (
	"gonum.org/v1/gonum/stat/distuv"
)

func main() {
	d.Fit(data, weights)
}
```

### What did you expect to happen?
According to the [wiki](https://en.wikipedia.org/wiki/Laplace_distribution) the scaling factor should be 
b=(1/n)sum|xi-m| 
### What actually happened?
the method is implemented with 
```golang 
// The scale parameter is the average absolute distance
// between the sample and the mean
absError := stat.MomentAbout(1, samples, l.Mu, weights)
```
However the moment is computed with
```golang
m += weights[i] * math.Pow(v-mean, moment)
```
Because we are at moment 1 so `Pow` is just `v-mean` hence m converge to zero. 

### What version of Go and Gonum are you using?
go version go1.11.1 darwin/amd64
513db5a0a3aa4236f0ace2604a014c157876c32e

### Does this issue reproduce with the current master?

Yes

I understand that the moment is correctly computed. The Fit function should be updated. If its of interest I can submit a patch.

### PR DESCRIPTION
**Pull request by mestorlx**
Fixed issue when weights are nil.
Fix Fit function
This PR closes #638 closes #639

### CODE REVIEW

**Comment by mestorlx** [comment]
@kortschak it was indeed ignoring for the scale parameter sorry about that. Fixed it and added a Test

File: stat/distuv/laplace_test.go [hunk 1]
```diff
@@ -56,3 +56,30 @@ func TestLaplaceProb(t *testing.T) {
...
+			nil, 1, 3},
+		{[]float64{10, 1, 1},
+			[]float64{10, 10, 10}, 1, 3},
+		{[]float64{10, 1, 1},
+			[]float64{0, 1, 1}, 1, 0},
+	}
+	for i, c := range cases {
+		d := Laplace{}
+		d.Fit(c.samples, c.weights)
+		compareNumbers(t, c.mu, d.Mu, i, "Location")
```
Comment (kortschak) [inline stat/distuv/laplace_test.go]:
```
		if !floats.EqualWithinAbsOrRel(d.Mu, test.wantMu, tol, tol) {
			t.Errorf("unexpected location result for test %d: got:%f, want:%f", i, d.Mu, test.wantMu)
		}
		if !floats.EqualWithinAbsOrRel(d.Scale, test.wantScale, tol, tol) {
			t.Errorf("unexpected scale result for test %d: got:%f, want:%f", i, d.Scale, test.wantScale)
		}
```

File: stat/distuv/laplace_test.go [hunk 2]
```diff
@@ -56,3 +56,30 @@ func TestLaplaceProb(t *testing.T) {
 	}
 	testDistributionProbs(t, Laplace{Mu: 0, Scale: 1}, "Laplace", pts)
 }
+
+func compareNumbers(t *testing.T, expected float64, got float64, i int, test string) {
```
Comment (kortschak) [inline stat/distuv/laplace_test.go]: Delete this.

File: stat/distuv/laplace_test.go [hunk 3]
```diff
@@ -56,3 +56,30 @@ func TestLaplaceProb(t *testing.T) {
...
+	if math.Abs(float64(expected-got)) > 1e-4 {
+		t.Errorf(test+" was incorrect, got: %f, expected: %f for set:%d", got, expected, i)
+	}
+}
+func TestLaplaceFit(t *testing.T) {
+	cases := []struct {
```
Comment (kortschak) [inline stat/distuv/laplace_test.go]:
```
	cases := []struct {
		samples   []float64
		weights   []float64
		wantMu    float64
		wantScale float64
	}{
		{
			samples:   []float64{10, 1, 1},
			weights:   nil,
			wantMu:    1,
			wantScale: 3,
		},
...
```

File: stat/distuv/laplace_test.go [hunk 4]
```diff
@@ -56,3 +56,30 @@ func TestLaplaceProb(t *testing.T) {
...
+		b       float64
+	}{
+		{[]float64{10, 1, 1},
+			nil, 1, 3},
+		{[]float64{10, 1, 1},
+			[]float64{10, 10, 10}, 1, 3},
+		{[]float64{10, 1, 1},
+			[]float64{0, 1, 1}, 1, 0},
+	}
+	for i, c := range cases {
```
Comment (kortschak) [inline stat/distuv/laplace_test.go]: `for i, test := range cases {`

**Comment by mestorlx** [comment]
@kortschak done.
Thanks for the review.

File: stat/distuv/laplace_test.go [hunk 5]
```diff
@@ -5,6 +5,7 @@
 package distuv
 
 import (
+	"gonum.org/v1/gonum/floats"
```
Comment (kortschak) [inline stat/distuv/laplace_test.go]: Can you run goimports here.

Comment (mestorlx) [reply stat/distuv/laplace_test.go]: done!

**Comment by kortschak** [comment]
Leaving for @btracey for the approval.

File: stat/distuv/laplace.go [hunk 6]
```diff
@@ -82,13 +82,20 @@ func (l *Laplace) Fit(samples, weights []float64) {
...
 
-	sumWeights := floats.Sum(weights)
-
 	// The scale parameter is the average absolute distance
 	// between the sample and the mean
-	absError := stat.MomentAbout(1, samples, l.Mu, weights)
-
-	l.Scale = absError / sumWeights
+	var absError float64
+	if weights != nil {
```
Comment (btracey) [inline stat/distuv/laplace.go]: Could you change this to sum up the weights as we pass through? That way we only do one pass through the weights slice.

File: stat/distuv/laplace_test.go [hunk 7]
```diff
@@ -56,3 +58,40 @@ func TestLaplaceProb(t *testing.T) {
...
+	for i, test := range cases {
+		d := Laplace{}
+		d.Fit(test.samples, test.weights)
+		if !floats.EqualWithinAbsOrRel(d.Mu, test.wantMu, 1e-10, 1e-10) {
+			t.Errorf("unexpected location result for test %d: got:%f, want:%f", i, d.Mu, test.wantMu)
+		}
+		if !floats.EqualWithinAbsOrRel(d.Scale, test.wantScale, 1e-10, 1e-10) {
+			t.Errorf("unexpected scale result for test %d: got:%f, want:%f", i, d.Scale, test.wantScale)
+		}
+	}
```
Comment (btracey) [inline stat/distuv/laplace_test.go]: Could you also add a randomized test? Something like

```
nSamples := 10000
src := rand.New(rand.NewSource(1))
l := Laplace {
    Mu: 3,
    Scale: 5,
    Src: src,
}
samples := make([]float64, nSamples)
for i := range samples {
    samples[i] = l.Rand()
}
l2 := Laplace{}
l2.Fit(samples, nil)
// Compare mu and scale to within 1e-2 or 1e-3
```

File: stat/distuv/laplace.go (hunk 6 above)
Comment (mestorlx) [reply stat/distuv/laplace.go]: Done

File: stat/distuv/laplace_test.go (hunk 7 above)
Comment (mestorlx) [reply stat/distuv/laplace_test.go]: done.

**Comment by btracey** [comment]
The change LGTM. Sorry, did you sign the Authors and Contributors?

**Comment by mestorlx** [comment]
Done

**Pull request merged**